linter_driver_rustc/src/ast/rustc.rs

```diff
@@ -26,7 +24,61 @@ impl<'ast, 'tcx> std::fmt::Debug for RustcContext<'ast, 'tcx> {
     }
 }
 
-impl<'ast, 'tcx> DriverContext<'ast> for RustcContext<'ast, 'tcx> {}
+fn to_leaked_rustc_lint(lint: &Lint) -> &'static RustcLint {
+    Box::leak(Box::new(RustcLint {
+        name: lint.name,
+        default_level: match lint.default_level {
+            Level::Allow => RustcLevel::Allow,
+            Level::Warn => RustcLevel::Warn,
+            Level::Deny => RustcLevel::Deny,
+            Level::Forbid => RustcLevel::Forbid,
+            _ => unreachable!("added variant to lint::Level"),
+        },
+        desc: lint.desc,
+        edition_lint_opts: None,
+        report_in_external_macro: match lint.report_in_macro {
+            MacroReport::No | MacroReport::Local => false,
+            MacroReport::All => true,
+            _ => unreachable!("added variant to lint::MacroReport"),
+        },
+        future_incompatible: None,
+        is_plugin: true,
+        feature_gate: None,
+        crate_level_only: false,
+    }))
+}
+
+impl<'ast, 'tcx> DriverContext<'ast> for RustcContext<'ast, 'tcx> {
+    fn warn(&self, s: &str, lint: &Lint) {
+        self.lint_ctx.lint(to_leaked_rustc_lint(lint), |diag| {
+            let mut diag = diag.build(s);
+            diag.emit();
+        });
+    }
+
+    fn warn_span(&self, s: &str, lint: &Lint, sp: &dyn Span<'_>) {
+        // Safety:
+        //
+        // Clearly this is probably not ideal but I did find this (answered by people much more
+        // knowledgeable about lifetime/unsafe/transmute)
+        // https://users.rust-lang.org/t/solved-transmute-between-trait-objects/13995/6
+        //
+        // In regard to the leak/forget, since we can keep this within the `'tcx` - `'ast` lifetime I don't
+        // think we have to. We aren't returning a `dyn Trait + 'static` like if we were actually using the
+        // `dyn Any` or the `dyn Span<'_>`. We can't use the `Any::downcast_ref` machinery here since we
+        // have a non `'static` trait object in `Span` (at least I couldn't get it to work)
+        let down_span: &RustcSpan<'ast, 'tcx> = unsafe {
+            let sp_ptr = sp as *const _ as *const (*mut (), *mut ());
+            &*(sp_ptr as *const dyn std::any::Any as *const _)
+        };
+
+        self.lint_ctx
+            .struct_span_lint(to_leaked_rustc_lint(lint), down_span.span, |diag| {
```

Comment:
This `to_leaked_rustc_lint()` call means that every time this function is called, a new lint instance is created and leaked. This is a bit problematic IMO. Another problem is, that rustc internally uses the memory address of lints for comparison, since these are static.

Ideally, we would create these lint instances in [linter_driver_rustc/src/main.rs](https://github.com/rust-linting/rust-linting/blob/1ed2c0a96ef282bf65f568bd18fc43d3faf039ab/linter_driver_rustc/src/main.rs#L40-L45) so that rustc can track the lint level. The issue with this is, that we would then have to somehow store all created lints and later save them in `RustcContext` and use them for emission. This in turn requires the lint modules to be loaded at that point as well. I've tried to do that, but ran into lifetime problems...

As an intermediate solution, I think that we should add a `HashMap` to `RustcContext` that can map from our lint instances to Rustc's leaked lint instances. For the emission, we would then only create a new instance if it's the first lint trigger and otherwise reuse the already create one.

Does this make sense? :upside_down_face:

Reply:
Well I had to also wrap the FxHashMap in a RefCell since we call it from a method that is `&self` which I think we should keep it that way. Eventually maybe this will be `Arc<Mutex<...>>` and we can run the lints in parallel and just merge the outputs nicely somehow keeping track of ordering :shrug:

anyways I'm pretty sure this is what you meant, good catch this would have been a LOT of leaking oops haha.